urdf/src/model.cpp

```diff
@@ -124,50 +145,35 @@ bool Model::initXml(TiXmlElement * robot_xml)
   return Model::initString(ss.str());
 }
 
-bool Model::initString(const std::string & xml_string)
+bool Model::initString(const std::string & data)
 {
   urdf::ModelInterfaceSharedPtr model;
 
-  // necessary for COLLADA compatibility
-  if (IsColladaData(xml_string)) {
-    fprintf(stderr, "Parsing robot collada xml string is not yet supported.\n");
-    return false;
-    /*
-    ROS_DEBUG("Parsing robot collada xml string");
-
-    static boost::mutex PARSER_PLUGIN_LOCK;
-    static boost::scoped_ptr<pluginlib::ClassLoader<urdf::URDFParser> > PARSER_PLUGIN_LOADER;
-    boost::mutex::scoped_lock _(PARSER_PLUGIN_LOCK);
-
-    try
-    {
-      if (!PARSER_PLUGIN_LOADER)
-        PARSER_PLUGIN_LOADER.reset(new pluginlib::ClassLoader<urdf::URDFParser>("urdf_parser_plugin", "urdf::URDFParser"));
-     const std::vector<std::string> &classes = PARSER_PLUGIN_LOADER->getDeclaredClasses();
-     bool found = false;
-     for (std::size_t i = 0 ; i < classes.size() ; ++i)
-       if (classes[i].find("urdf/ColladaURDFParser") != std::string::npos)
-     {
-       boost::shared_ptr<urdf::URDFParser> instance = PARSER_PLUGIN_LOADER->createInstance(classes[i]);
-       if (instance)
-         model = instance->parse(xml_string);
-       found = true;
-       break;
-     }
-     if (!found)
-       ROS_ERROR_STREAM("No URDF parser plugin found for Collada files. Did you install the corresponding package?");
-   }
-   catch(pluginlib::PluginlibException& ex)
-   {
-     ROS_ERROR_STREAM("Exception while creating planning plugin loader " << ex.what() << ". Will not parse Collada file.");
-   }
+  size_t best_score = std::numeric_limits<size_t>::max();
+  pluginlib::UniquePtr<urdf::URDFParser> best_plugin;
+  std::string best_plugin_name;
+
+  // Figure out what plugins might handle this format
+  for (const std::string & plugin_name : impl_->loader_.getDeclaredClasses()) {
+    pluginlib::UniquePtr<urdf::URDFParser> plugin_instance =
+      impl_->loader_.createUniqueInstance(plugin_name);
+    if (plugin_instance) {
+      size_t score = plugin_instance->might_handle(data);
+      if (score < best_score) {
+        best_score = score;
+        best_plugin = std::move(plugin_instance);
+        best_plugin_name = plugin_name;
+      }
+    }
```

Comment:
What are the situations that could lead to `plugin_instance` being `nullptr`?  Should we warn on those instances?

Reply:
Looks like [never](https://github.com/ros/class_loader/blob/8864be372a83921f15906ae80752be9d6a300b81/include/class_loader/class_loader.hpp#L302), but it [might raise an exception](https://github.com/ros/pluginlib/blob/b076db77bee291f67d02540c7eca1db02d0b0ccb/pluginlib/include/pluginlib/class_loader_imp.hpp#L201).

Handle pluginlib exception in 33ad016

Reply:
I will point out that those asserts (both down in pluginlib and the new one you added here) will be compiled out if you build in `Release` mode.  In that mode, if `createPluginInstance` returns a nullptr, and then this plugin will crash on the call to `plugin_handle->might_handle()`.

That being said, I think we also need the handling for the exception that you've added.  So I'd suggest a hybrid between the old code and the new code; wrap the thing in a `try...catch` block, and also make sure to check for nullptr.

Reply:
Avoids crashing by dereferencing a nullptr in 3841ea4